File: Test/EscortDecomposingSpec.m
#import "Kiwi.h"
#import "NSDate+Escort.h"
#import "NSDate+AZDateBuilder.h"

@interface NSDate ()
+ (NSCalendar *)AZ_currentCalendar;
@end

SPEC_BEGIN(EscortDecomposingSpec)
    registerMatchers(@"AZ");// NSDate Custom Matcher
    describe(@"-nearestHour", ^{
        context(@"when 10:00:00", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @0,
                    AZ_DateUnit.second : @0,
                }];
            });
            it(@"should return 10", ^{
                NSInteger nearestHour = [currentDate nearestHour];
                [[theValue(nearestHour) should] equal:theValue(10)];
            });
        });
        context(@"when 10:29:59", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @29,
                    AZ_DateUnit.second : @59,
                }];
            });
            it(@"should return 10", ^{
                NSInteger nearestHour = [currentDate nearestHour];
                [[theValue(nearestHour) should] equal:theValue(10)];
            });
        });
        context(@"when 10:30:00", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @10,
                    AZ_DateUnit.minute : @30,
                    AZ_DateUnit.second : @0,
                }];
            });
            it(@"should return 11", ^{
                NSInteger nearestHour = [currentDate nearestHour];
                [[theValue(nearestHour) should] equal:theValue(11)];
            });
        });
    });
    describe(@"-hour", ^{
        context(@"when the date is 01:02:03", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @1,
                    AZ_DateUnit.minute : @2,
                    AZ_DateUnit.second : @3,
                }];
            });
            it(@"should return 1", ^{
                NSInteger hour = [currentDate hour];
                [[theValue(hour) should] equal:theValue(1)];
            });
        });
    });
    describe(@"-minute", ^{
        context(@"when the date is 01:02:03", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @1,
                    AZ_DateUnit.minute : @2,
                    AZ_DateUnit.second : @3,
                }];
            });
            it(@"should return 2", ^{
                NSInteger minute = [currentDate minute];
                [[theValue(minute) should] equal:theValue(2)];
            });
        });
    });
    describe(@"-seconds", ^{
        context(@"when the date is 01:02:03", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                    AZ_DateUnit.hour : @1,
                    AZ_DateUnit.minute : @2,
                    AZ_DateUnit.second : @3,
                }];
            });
            it(@"should return 3", ^{
                NSInteger seconds = [currentDate seconds];
                [[theValue(seconds) should] equal:theValue(3)];
            });
        });
    });
    describe(@"-day", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];
            });
            it(@"should return 10", ^{
                NSInteger day = [currentDate day];
                [[theValue(day) should] equal:theValue(10)];
            });
        });
    });
    describe(@"-month", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];
            });
            it(@"should return 10", ^{
                NSInteger month = [currentDate month];
                [[theValue(month) should] equal:theValue(10)];
            });
        });
    });
    describe(@"-week", ^{
        context(@"when the date 2010-01-01", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @1,
                    AZ_DateUnit.day : @1,
                }];
            });
            it(@"should return 1", ^{
                NSInteger weekValue = [currentDate week];
                [[theValue(weekValue) should] equal:theValue(1)];
            });
        });
    });
    describe(@"-weekday", ^{
        // unfortunately NSCalendar$setFirstWeekday is not configurable...
        context(@"when the date 2010-01-01(Fri)", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @1,
                    AZ_DateUnit.day : @1,
                }];
            });
            it(@"should return 6", ^{
                NSInteger weekdayValue = [currentDate weekday];
                [[theValue(weekdayValue) should] equal:theValue(6)];
            });
        });
    });

    describe(@"-nthWeekday", ^{
        context(@"when the date 2010-10-10(2th weekday)", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];
            });
            it(@"should return 2", ^{
                NSInteger nthWeekday = [currentDate nthWeekday];
                [[theValue(nthWeekday) should] equal:theValue(2)];
            });
        });
    });
    describe(@"-firstDayOfWeekday", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(10)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 4", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(4)];
                });
            });
        });
        context(@"when the date 2010-10-11", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @11,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(10)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 11", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(11)];
                });
            });
        });
        context(@"when the date 2010-10-15", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @15,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(10)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 11", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(11)];
                });
            });
        });
        context(@"when the date 2010-10-16", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @16,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(10)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger firstDayOfWeekday = [currentDate firstDayOfWeekday];
                    [[theValue(firstDayOfWeekday) should] equal:theValue(11)];
                });
            });
        });
    });
    describe(@"-lastDayOfWeekday", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(16)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 4", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(10)];
                });
            });
        });
        context(@"when the date 2010-10-11", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @11,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(16)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 11", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(17)];
                });
            });
        });
        context(@"when the date 2010-10-15", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @15,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(16)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 11", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(17)];
                });
            });
        });
        context(@"when the date 2010-10-16", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @16,// weekday is 1
                }];
            });
            context(@"begining of sunday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 1;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(16)];
                });
            });
            context(@"begining of monday for weekady", ^{
                beforeEach(^{
                    NSCalendar *beginingOfMondayCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
                    beginingOfMondayCalendar.firstWeekday = 2;
                    [NSDate stub:@selector(AZ_currentCalendar) andReturn:beginingOfMondayCalendar];
                });
                it(@"should return 10", ^{
                    NSInteger lastDayOfWeekday = [currentDate lastDayOfWeekday];
                    [[theValue(lastDayOfWeekday) should] equal:theValue(17)];
                });
            });
        });
    });
    describe(@"-year", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];
            });
            it(@"should return 2010", ^{
                NSInteger yearValue = [currentDate year];

                NSCalendar *c = [NSCalendar currentCalendar];
                if ([c.calendarIdentifier isEqualToString:NSCalendarIdentifierGregorian]) {
                    [[theValue(yearValue) should] equal:theValue(2010)];
                } else if ([c.calendarIdentifier isEqualToString:NSCalendarIdentifierJapanese]) {
                    [[theValue(yearValue) should] equal:theValue(22)];
                } else {
                    [[theValue(yearValue) shouldNot] equal:theValue(2010)];
                    [[theValue(yearValue) shouldNot] equal:theValue(22)];
                }
            });
        });
        context(@"when the calendar is not Gregorian", ^{
            __block NSCalendar *jaCalendar;
            beforeEach(^{
                // +currentCalendar overwrite NSCalendarIdentifierJapanese.
                jaCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierJapanese];
                [NSDate stub:@selector(AZ_currentCalendar) andReturn:jaCalendar];
            });
            context(@"simple compere", ^{
                __block NSDate *currentDate;
                __block NSDate *expectedDate;
                beforeEach(^{
                    expectedDate = [NSDate AZ_dateByUnit:@{
                        AZ_DateUnit.year : @2010,
                        AZ_DateUnit.month : @10,
                        AZ_DateUnit.day : @10,
                    }];

                    NSDateComponents *dateComponents = [[NSDateComponents alloc] init];
                    dateComponents.year = 22;
                    dateComponents.month = 10;
                    dateComponents.day = 10;
                    currentDate = [jaCalendar dateFromComponents:dateComponents];
                });
                it(@"should return 22", ^{
                    NSInteger yearValue = [currentDate year];
                    [[theValue(yearValue) should] equal:theValue(22)];
                });
                it(@"the date equal to 2010-10-10", ^{
                    [[currentDate should] equal:expectedDate];
                });
            });
            context(@"when Shouwa era", ^{
                __block NSDate *currentDate;
                beforeEach(^{
                    currentDate = [NSDate AZ_dateByUnit:@{
                        AZ_DateUnit.year : @1988,
                        AZ_DateUnit.month : @10,
                        AZ_DateUnit.day : @10,
                    }];
                });
                it(@"should return 63", ^{
                    NSInteger yearValue = [currentDate year];
                    [[theValue(yearValue) should] equal:theValue(63)];
                });
            });
        });
    });

    describe(@"-gregorianYear", ^{
        context(@"when the date 2010-10-10", ^{
            __block NSDate *currentDate;
            beforeEach(^{
                currentDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];
            });
            it(@"should return 2010", ^{
                NSInteger yearValue = [currentDate year];

                NSCalendar *c = [NSCalendar currentCalendar];
                if ([c.calendarIdentifier isEqualToString:NSCalendarIdentifierGregorian]) {
                    [[theValue(yearValue) should] equal:theValue(2010)];
                } else if ([c.calendarIdentifier isEqualToString:NSCalendarIdentifierJapanese]) {
                    [[theValue(yearValue) should] equal:theValue(22)];
                } else {
                    [[theValue(yearValue) shouldNot] equal:theValue(2010)];
                    [[theValue(yearValue) shouldNot] equal:theValue(22)];
                }
            });
        });
        context(@"when the date'calendar is not Gregorian", ^{
            __block NSDate *currentDate;
            __block NSDate *expectedDate;
            beforeEach(^{
                expectedDate = [NSDate AZ_dateByUnit:@{
                    AZ_DateUnit.year : @2010,
                    AZ_DateUnit.month : @10,
                    AZ_DateUnit.day : @10,
                }];

                NSCalendar *jaCalendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierJapanese];
                NSDateComponents *dateComponents = [[NSDateComponents alloc] init];
                dateComponents.year = 1;
                dateComponents.month = 10;
                dateComponents.day = 10;
                currentDate = [jaCalendar dateFromComponents:dateComponents];
                // +currentCalendar overwrite NSCalendarIdentifierJapanese.
                [NSDate stub:@selector(AZ_currentCalendar) andReturn:jaCalendar];
            });
            it(@"should return 1989", ^{
                NSInteger yearValue = [currentDate gregorianYear];
                [[theValue(yearValue) should] equal:theValue(1989)];
            });
            it(@"the date equal to 2010-10-10", ^{
                [[theValue(currentDate) shouldNot] equal:expectedDate];
            });
        });
    });

SPEC_END

File: SwiftTest/EscortDecomposingSpec.swift
import XCTest
import Quick
import AZDateBuilder

class EscortDecomposingSpec: QuickSpec {
    override func spec() {
        describe("era") {
            context("when 2010-10-10") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 15,
                    ])
                it("gregorian calendar should return 1") {
                    Date.identifier = .gregorian
                    XCTAssertEqual(currentDate.era, 1)
                }
                it("japanese calendar should return 1") {
                    Date.identifier = .japanese
                    XCTAssertEqual(currentDate.era, 235)
                }
            }
            context("when Showa era") {
                let currentDate = Date.date(by: [
                    .year: 1988,
                    .month: 10,
                    .day: 11,
                    ])
                it("japanese calendar should return 63") {
                    Date.identifier = .japanese
                    XCTAssertEqual(currentDate.era, 234)
                }
            }
        }
        describe("year") {
            context("when 2010-10-10") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 15,
                    ])
                it("gregorian calendar should return 2010") {
                    Date.identifier = .gregorian
                    XCTAssertEqual(currentDate.year, 2010)
                }
                it("japanese calendar should return 22") {
                    Date.identifier = .japanese
                    XCTAssertEqual(currentDate.year, 22)
                }
            }
            context("when Showa era") {
                let currentDate = Date.date(by: [
                    .year: 1988,
                    .month: 10,
                    .day: 11,
                    ])
                it("japanese calendar should return 63") {
                    Date.identifier = .japanese
                    XCTAssertEqual(currentDate.year, 63)
                }
            }
        }
        describe("month") {
            it("when 2010-10-11 should retuen 11") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    ])
                XCTAssertEqual(currentDate.month, 10)
            }
        }
        describe("day") {
            it("when 2010-10-11 should retuen 11") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    ])
                XCTAssertEqual(currentDate.day, 11)
            }
        }
        describe("minute") {
            it("when 01:02:03 should retuen 2") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    .hour: 1,
                    .minute: 2,
                    .second: 3,
                    ])
                XCTAssertEqual(currentDate.minute, 2)
                
            }
        }
        describe("second") {
            it("when 01:02:03 should retuen 3") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    .hour: 1,
                    .minute: 2,
                    .second: 3,
                    ])
                XCTAssertEqual(currentDate.second, 3)
            }
        }
        describe("weekday") {
            it("when 2010-01-01 should retuen 6") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 1,
                    .day: 1,
                    ])
                XCTAssertEqual(currentDate.weekday, 6)
            }
            it("when 2010-10-11 should retuen 2") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    ])
                XCTAssertEqual(currentDate.weekday, 2)
            }
        }
        describe("weekdayOrdinal") {
            it("when 2010-01-01 should retuen 1") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 1,
                    .day: 1,
                    ])
                XCTAssertEqual(currentDate.weekdayOrdinal, 1)
            }
            it("when 2010-10-07 should retuen 1") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 7,
                    ])
                XCTAssertEqual(currentDate.weekdayOrdinal, 1)
            }
            it("when 2010-10-08 should retuen 2") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 8,
                    ])
                XCTAssertEqual(currentDate.weekdayOrdinal, 2)
            }
            it("when 2010-10-14 should retuen 2") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 14,
                    ])
                XCTAssertEqual(currentDate.weekdayOrdinal, 2)
            }
            it("when 2010-10-15 should retuen 3") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 15,
                    ])
                XCTAssertEqual(currentDate.weekdayOrdinal, 3)
            }
        }
        describe("quarter") {
            it("when 2010-01 should retuen 0") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 1,
                    ])
                XCTAssertEqual(currentDate.quarter, 0)
            }
        }
        describe("weekOfMonth") {
            it("when 2010-10-09 should retuen 2") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 9,
                    ])
                XCTAssertEqual(currentDate.weekOfMonth, 2)
            }
            it("when 2010-10-10 should retuen 3") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 10,
                    ])
                XCTAssertEqual(currentDate.weekOfMonth, 3)
            }
            it("when 2010-10-16 should retuen 3") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 16,
                    ])
                XCTAssertEqual(currentDate.weekOfMonth, 3)
            }
            it("when 2010-10-17 should retuen 4") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 17,
                    ])
                XCTAssertEqual(currentDate.weekOfMonth, 4)
            }
            it("when 2010-10-31 should retuen 4") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 31,
                    ])
                XCTAssertEqual(currentDate.weekOfMonth, 6)
            }
        }
        describe("weekOfYear") {
            it("when 2010-10-09 should retuen 41") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 9,
                    ])
                XCTAssertEqual(currentDate.weekOfYear, 41)
            }
            it("when 2010-10-10 should retuen 42") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 10,
                    ])
                XCTAssertEqual(currentDate.weekOfYear, 42)
            }
            it("when 2010-10-16 should retuen 42") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 16,
                    ])
                XCTAssertEqual(currentDate.weekOfYear, 42)
            }
            it("when 2010-10-17 should retuen 43") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 17,
                    ])
                XCTAssertEqual(currentDate.weekOfYear, 43)
            }
            it("when 2010-12-26 should retuen 1") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 12,
                    .day: 26,
                    ])
                XCTAssertEqual(currentDate.weekOfYear, 1)
            }
        }
        describe("yearForWeekOfYear") {
            it("when 2009-12-26 should retuen 2009") {
                let currentDate = Date.date(by: [
                    .year: 2009,
                    .month: 12,
                    .day: 26,
                    ])
                XCTAssertEqual(currentDate.yearForWeekOfYear, 2009)
            }
            it("when 2009-12-27 should retuen 2010") {
                let currentDate = Date.date(by: [
                    .year: 2009,
                    .month: 12,
                    .day: 27,
                    ])
                XCTAssertEqual(currentDate.yearForWeekOfYear, 2010)
            }
            it("when 2010-12-25 should retuen 2010") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 12,
                    .day: 25,
                    ])
                XCTAssertEqual(currentDate.yearForWeekOfYear, 2010)
            }
            it("when 2010-12-26 should retuen 2011") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 12,
                    .day: 26,
                    ])
                XCTAssertEqual(currentDate.yearForWeekOfYear, 2011)
            }
        }
        describe("nanosecond") {
            it("when 01:02:03.059 should retuen 59") {
                let currentDate = Date.date(by: [
                    .year: 2010,
                    .month: 10,
                    .day: 11,
                    .hour: 1,
                    .minute: 2,
                    .second: 3,
                    .nanosecond: 59,
                    ])
                XCTAssertEqual(currentDate.nanosecond, 59)
            }
        }
    }
}
